### ISSUE TITLE
DASH/CENC with ClearKey

### ISSUE THREAD

**Issue by andreyhammer** [kaltura/nginx-vod-module#547]
Hello!

I want to use own drm license key based on Clear Key. It is possible to use your module  with  clear key?
In docs I see - "DASH/CENC with PlayReady & Widevine PSSH together" but no any about clear key.

And thank you for your software!

**Comment by erankor** [comment]
Clear key is not supported, do you have some working stream that uses it? If you do, I may add support for it.
I tried to look for a sample at one point and didn't find any, there was only one from Azure media services, but it played only on their player (didn't play on both Shaka/Dashif IIRC)

**Comment by andreyhammer** [comment]
No, I don't have working stream. 
I have box without internet acces with nginx  and unencrypted content onboard and  I want to get on-fly DASH/CENC encription with clear key.  This box for onboard entertainment and need to prevent content leaking,
Ok, thank you for your anwser.  Will check other ways.

**Comment by erankor** [comment]
Ok, just to clarify - I'll be happy to add this feature, but up until now, I didn't see any such stream working on Shaka player (only found streams that work on Azure media player). If anyone here has a sample clear keys stream working on Shaka, please share it.
Regarding alternative solutions, I don't know which devices you are targeting, but you may be able to use HLS AES128 with the hls.js player.

**Comment by erankor** [comment]
Closing this for now, if anyone has a clear keys sample, please reopen

**Comment by xylomo** [comment]
Hi erankor,
First, thank you for your work.

Also, on the Shaka Player Demo page (https://shaka-player-demo.appspot.com) there is a clear key asset you can use. It is called Angel One (multicodec, multilingual, ClearKey server). The manifest can be found here: https://storage.googleapis.com/shaka-demo-assets/angel-one-clearkey/dash.mpd. Will this help in supporting ClearKey? I was also looking to see if this can be implemented.

**Comment by erankor** [comment]
Thanks @xylomo, I followed this sample and got a clear key stream playing in Shaka:
Manifest: http://lbd.kaltura.com:8001/local/edash/content/ff.mp4/manifest.mpd
License server: http://lbd.kaltura.com:8001/dash_clear_key.php

When using the Shaka demo it doesn't support by default the use of clear key with a custom license server. I patched asset_section.js locally in Chrome developer tools to work around that. Added `'org.w3.clearkey': licenseServer` here - https://github.com/google/shaka-player/blob/master/demo/asset_section.js#L150.

Check out this pull request - https://github.com/kaltura/nginx-vod-module/pull/588
It includes a sample php backend for supporting clear key, this script generates both the DRM JSON that nginx-vod-module requires, and also answers license requests from users.
I may add built-in support for this at some point (similarly to how HLS-AES works), but for now it's written using an external server. 
Note that you also need the nginx-vod code change in the mentioned pull for this to work.
